### PR TITLE
Implement batch mode via the CLI.

### PR DESCRIPTION
**Pull request by ColinArenz**
- Run multiple queries simultaneously, outputting the results to separate table files.
- Index arbitrary sentence corpora as opposed to the existing pre-processed set.
- Persist new corpus index between multiple executions.

### CODE REVIEW

**Comment by ColinArenz** [comment]
Adding @chmmd as FYI.

**Comment by rodneykinney** [comment]
Rock on!

**Comment by dirkgr** [comment]
Tests failed?

File: src/main/resources/application.conf [hunk 1]
```diff
@@ -33,19 +33,19 @@ IkeToolWebapp = {
...
+    // NOTE: Before running locally, uncomment the below lines and specify values for keys `url` to point to your PostgresSql JDBC
     // link, `user` and `password`.
     // AI2 Internal users: use the database in the `test` deploy environment and the same username. You can get the password
     // from this file in the ops-keystore in S3.
-    // url =
-    // user =
-    // password =
+    url = "jdbc:postgresql://"${POSTGRES_DB}":5432/okcorpus"
+    user = "okcorpus"
+    password = ${POSTGRES_PASSWORD}
```
Comment (dirkgr) [inline src/main/resources/application.conf]: Why did you remove this? Can you still deploy to the IKE production server without this?

**Comment by ColinArenz** [comment]
The tests failed on code formatting, but the whole project suffers from formatting issues.  The next change will be pretty noisy (pure formatting), but can safely be ignored.

File: src/main/resources/application.conf (hunk 1 above)
Comment (ColinArenz) [inline src/main/resources/application.conf]: This isn't removed.  It's commented out in master, and needs to be manually updated so we don't publish our passwords.  Rodney wrote this handy config so we can just set a couple of environment variables, but I can revert it if you feel strongly.  @rodneykinney: Thoughts?

Comment (dirkgr) [inline src/main/resources/application.conf]: Sorry, I was reading this backwards. Brain fart.

File: src/main/scala/org/allenai/ike/IkeBatchSearch.scala [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+    }
+
+    /** Split a pattern line into a pattern name and table name, as a tuple
+      *
+      * @param patternLine the pattern line to split
+      * @return a tuple with (pattern name, table name)
+      */
+    def splitPattern(patternLine: String): Option[(String, String)] = {
+      patternLine.split("\t").map(_.trim) match {
+        case Array(patternName, tableName, _*) => Some((patternName, tableName))
```
Comment (dirkgr) [inline src/main/scala/org/allenai/ike/IkeBatchSearch.scala]: Third column and onwards are comments?

File: src/main/scala/org/allenai/ike/IkeBatchSearch.scala [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+        case _ => None
+      }
+    }
+
+    val sparkConf = new SparkConf()
+      .setAppName(this.getClass.getSimpleName)
+      .setMaster("local[*]")
```
Comment (dirkgr) [inline src/main/scala/org/allenai/ike/IkeBatchSearch.scala]: I'm assuming this is a default and you override this to run on a cluster later?

File: src/main/scala/org/allenai/ike/IkeBatchSearch.scala [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+      .set("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
+      .set("spark.kryo.registrator", "org.allenai.ike.IkeKryoRegistrator")
```
Comment (dirkgr) [inline src/main/scala/org/allenai/ike/IkeBatchSearch.scala]: Does using Kryo make a difference for performance? I've never seen some solid numbers on that. Would be interested.

File: src/main/scala/org/allenai/ike/IkeBatchSearch.scala [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+    val sparkContext = new SparkContext(sparkConf)
+
+    // Open the index directory or create a temp directory if one hasn't been provided.
+    val indexParentDir = batchSearchOptions.indexPath.map { path =>
+      val dir = new File(path)
+      dir.mkdir()
+      dir
+    }.getOrElse {
+      val tempDir = Files.createTempDirectory("ike").toFile
+      tempDir.deleteOnExit()
```
Comment (dirkgr) [inline src/main/scala/org/allenai/ike/IkeBatchSearch.scala]: You can't `deleteOnExit` for directories. I wrote a workaround for that once, but it's part of the datastore code, so not in a super accessible part of the code.

File: src/main/scala/org/allenai/ike/IkeBatchSearch.scala (hunk 2 above)
Comment (ColinArenz) [inline src/main/scala/org/allenai/ike/IkeBatchSearch.scala]: There shouldn't be a third column, so this will ignore everything after the second tab.  We can make this more strict by requiring exactly two elements in the array?

File: src/main/scala/org/allenai/ike/IkeBatchSearch.scala (hunk 3 above)
Comment (ColinArenz) [inline src/main/scala/org/allenai/ike/IkeBatchSearch.scala]: At present we only support local runs, but with some additional work we should be able to scale this out by saving all the files in HDFS.

File: src/main/scala/org/allenai/ike/IkeBatchSearch.scala (hunk 4 above)
Comment (ColinArenz) [inline src/main/scala/org/allenai/ike/IkeBatchSearch.scala]: Not on a local machine, since shuffles aren't bounded by network traffic.  On a remote cluster, this matters more.  However, the Spark documentation says Kryo should be the default; they just didn't make that change for back-compat reasons.

File: src/main/scala/org/allenai/ike/IkeBatchSearch.scala [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+      indexDirectories.map(_.getAbsolutePath)
+    } else {
+      // Load the text source and partition it according to ideal index size. The maximum sentences
+      // per partition were determined through several experimental runs. As indices get larger,
+      // indexing slows down. Since partitioning in our scenario is arbitrary and relevance doesn't
+      // matter, we should optimize strictly for speed.
+      val textLines = sparkContext.textFile(batchSearchOptions.textSource)
+      val sentences = textLines.repartition {
+        val maxSentencesPerPartition = 256 * 1024
+        val minPartitions = Runtime.getRuntime.availableProcessors
```
Comment (dirkgr) [inline src/main/scala/org/allenai/ike/IkeBatchSearch.scala]: This is the number of processors on the driver? That's not useful when running on a cluster.

File: src/main/scala/org/allenai/ike/IkeBatchSearch.scala [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+            }
+          }
+          indexer.close()
+
+          Seq(indexDir.getAbsolutePath).toIterator
+      }.collect
+    }
+
+    // For each index and pattern, search and dump the serialized results.
+    val patternResults: Seq[(String, String)] = indexPaths.toSeq.par.flatMap { indexPath =>
```
Comment (dirkgr) [inline src/main/scala/org/allenai/ike/IkeBatchSearch.scala]: We're using `par` and Spark parallelization in the same app?

File: src/main/scala/org/allenai/ike/IkeBatchSearch.scala [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+    }.sortBy(-_._2)
+
+    // Output the results for WebUI import to a flat file.
+    val webFile = new File(outputDirPath, tableName + "_webui.tsv")
+    Resource.using(new PrintWriter(webFile)) { writer =>
+      writer.println(columnHeaders.mkString("\t") + "\tlabel\tprovenance")
+      hitsWithCounts.toLocalIterator.foreach {
+        case (output, count) =>
+          // TODO: Create an 'unlabeled' category for the batch mode scenario.
+          // TODO: Save provenance.
```
Comment (dirkgr) [inline src/main/scala/org/allenai/ike/IkeBatchSearch.scala]: Nobody uses provenance.

File: src/main/scala/org/allenai/ike/IkeKryoRegistrator.scala [hunk 9]
```diff
@@ -0,0 +1,53 @@
+package org.allenai.ike
+
+import org.allenai.ike.patterns.NamedPattern
+
+import com.esotericsoftware.kryo.io.{ Input, Output }
+import com.esotericsoftware.kryo.{ Kryo, Serializer }
+import org.apache.spark.serializer.KryoRegistrator
+
+/** Helper to register custom serializers for Some and None. Since Kryo is a Java library, it
+  * doesn't provide default serializers for these classes.
```
Comment (dirkgr) [inline src/main/scala/org/allenai/ike/IkeKryoRegistrator.scala]: Chill is the name of the library that provides Kryo serializations for Scala types.

**Comment by dirkgr** [comment]
The main concern I have with this is that it doesn't know whether it wants to run in Spark or locally. If it's local-only, it needs no Spark at all. Everything could be done with `par`. If it needs to be distributed, it needs some work.

**Comment by ColinArenz** [comment]
I agree with the principle.  Our line of thinking follows:
- Spark gets us free parallel data parsing from file or S3, without loading it all into memory.
- We probably want to distribute this for larger corpora.  Even a few million sentences takes a while.  Starting with Spark on local makes for a relatively easy transition to clusters in a later change.

I can remove the Spark, but it will probably be a significant rework.

File: src/main/scala/org/allenai/ike/IkeBatchSearch.scala (hunk 6 above)
Comment (ColinArenz) [inline src/main/scala/org/allenai/ike/IkeBatchSearch.scala]: There's no way to get the available cores on a cluster that I'm aware of.  We could add a config value or a command line argument.  For very large corpora, this number doesn't matter.  For small corpora on a local machine, this gets max parallelism.  I can set it to an arbitrary value for now.

File: src/main/scala/org/allenai/ike/IkeBatchSearch.scala (hunk 7 above)
Comment (ColinArenz) [inline src/main/scala/org/allenai/ike/IkeBatchSearch.scala]: This portion can't be easily Sparkified due to the large number of things that are passed in (tables, etc).  If we rebuilt from the ground up it would be a possibility.

File: src/main/scala/org/allenai/ike/IkeBatchSearch.scala (hunk 8 above)
Comment (ColinArenz) [inline src/main/scala/org/allenai/ike/IkeBatchSearch.scala]: Removed.

File: src/main/scala/org/allenai/ike/IkeKryoRegistrator.scala (hunk 9 above)
Comment (ColinArenz) [inline src/main/scala/org/allenai/ike/IkeKryoRegistrator.scala]: Indeed, but that seems a bit heavy for this one class?

**Comment by rodneykinney** [comment]
Using `.par` is not enough because there is also a `groupBy` that causes an out-of-memory error in some cases.

File: src/main/scala/org/allenai/ike/IkeBatchSearch.scala (hunk 5 above)
Comment (ColinArenz) [inline src/main/scala/org/allenai/ike/IkeBatchSearch.scala]: Fixed using datastore's `TempCleanup`.

**Comment by dirkgr** [comment]
> Using .par is not enough because there is also a groupBy that causes an out-of-memory error in some cases.

You can give it more memory. Efficiency wise, it's competing with a whole spark cluster, right? That's a lot of memory.

To put it in Spark terms, you're doing some fairly heavy lifting on the driver, which you aren't supposed to do, according to Spark doctrine.

I realize it's hard to do a switch in either direction. If you want to merge anyways, I guess that's fine. Also, I can see how you would end up with this solution during a hackathon. But for production, that's a pretty big design problem.

File: src/main/resources/application.conf (hunk 1 above)
Comment (sbhaktha) [inline src/main/resources/application.conf]: This comes from either the `test.conf` or the `prod.conf`depending on the chosen deploy environment, from the `deploy` repo at deployment time. (That's 3 mentions of `deploy` in one sentence :-)). This stuff should not only be commented but the earlier contents should not be reachable via git history.

File: src/main/scala/org/allenai/ike/IkeBatchSearch.scala (hunk 2 above)
Comment (dirkgr) [inline src/main/scala/org/allenai/ike/IkeBatchSearch.scala]: Maybe we should standardize on a TSV/CSV parser, one that supports escaping and comments. But not in this PR :-)

File: src/main/scala/org/allenai/ike/IkeKryoRegistrator.scala (hunk 9 above)
Comment (dirkgr) [inline src/main/scala/org/allenai/ike/IkeKryoRegistrator.scala]: That's not how I would make that trade-off, but it's a reasonable reason.

I would go with Chill on autopilot. On reflection though, the reason is that otherwise, when you develop this more, you end up with situations where it works with a 2-tuple, but then you change the code a bit to use a 3-tuple, and suddenly it stops working at runtime, in a part of the code you know nothing about.

Like I said though, no big deal. Keeping the dependencies light has its own perks.

**Comment by rodneykinney** [comment]
This isn't running on a cluster; it's running in local mode. I think it's a fair use of Spark to do grouping, sorting, and aggregation on a single machine for datasets that don't fit in memory.

**Comment by ColinArenz** [comment]
I can't get these tests to pass, and it looks to be an issue with Semaphore.  I've even tried commenting out all the tests in the failing `TestQuerySuggester` and Semaphore still reports that test as "failed".  It sounds like we're pretty much out of ideas at this point, so I'm going to give up on this PR.

@sbhaktha: I've pushed all these changes to the `hackathon` branch so we at least have them somewhere.  Can you take a look when you're back?
